frontend/src/js/util/auth/authUploadWithProgress.ts

```diff
@@ -1,57 +1,152 @@
+  reject: (reason?: any) => void,
+  workspace?: WorkspaceUploadMetadata
+) => {
+  xhr.onloadend = () => {
+    handleResponseFromAuthRequest(
+      xhr.status,
+      xhr.getResponseHeader("X-Offer-Authorization")
+    );
+    if (xhr.status >= 200 && xhr.status < 300) {
+      resolve(xhr.response);
+    } else {
+      if (retryCount < RETRY_MAX_LIMIT) {
+        console.warn(
+          `retrying because request failed due to ${xhr.responseText} - status: ${xhr.status}, retry: ${retryCount} for url: ${url}`
+        );
+        retryRequest(
+          retryCount + 1,
+          xhr,
+          url,
+          file,
+          path,
+          uploadId,
+          resolve,
+          reject,
+          workspace
+        );
+      } else {
+        console.error(
+          `request failed due to ${xhr.responseText} - status: ${xhr.status}, retry: ${retryCount} for url: ${url}`
+        );
+        reject(`${xhr.status} - ${xhr.responseText}`);
+      }
+    }
+  };
+
+  console.log(`sending request: ${url} - retry: ${retryCount}`);
+  sendRequest(xhr, url, file, path, uploadId, workspace);
+};
+
+const retryRequest = (
+  retryCount: number,
+  xhr: XMLHttpRequest,
+  url: string,
+  file: File,
+  path: string,
+  uploadId: string,
+  resolve: (value: unknown) => void,
+  reject: (reason?: any) => void,
+  workspace?: WorkspaceUploadMetadata
+) => {
+  const limit = retryCount ? Math.pow(2, retryCount - 1) * 1000 : 0;
+  const pause = Math.random() * limit;
+  console.log(`delaying the next retry by ${pause / 1000} s - for url: ${url}`);
+
+  setTimeout(() => {
+    processRequest(
+      xhr,
+      url,
+      file,
+      path,
+      uploadId,
+      retryCount,
+      resolve,
+      reject,
+      workspace
+    );
+  }, pause);
+};
+
+const sendRequest = (
+  xhr: XMLHttpRequest,
+  url: string,
+  file: File,
+  path: string,
+  uploadId: string,
+  workspace?: WorkspaceUploadMetadata
+) => {
+  xhr.open("POST", url);
+
+  // this isn't nice, but I don't want to get bogged down typing all the redux stuff
+  const state = store.getState() as any;
+  xhr.setRequestHeader("Authorization", "Bearer " + state.auth.jwtToken);
+
+  xhr.setRequestHeader("Content-Type", file.type);
+  xhr.setRequestHeader("Content-Location", encodeURIComponent(path));
+  // TODO MRB: this is a lot of headers which is a bit silly and also potentially too 'custom'
+  // for the more intrusive firewalls of the world. Uploading as FormData is probably cleaner
+  // but we'd need to check it's still fine for big'ish files
+  xhr.setRequestHeader("X-PFI-Upload-Id", uploadId);
+  if (workspace) {
+    xhr.setRequestHeader("X-PFI-Workspace-Id", workspace.workspaceId);
+    xhr.setRequestHeader(
+      "X-PFI-Workspace-Parent-Node-Id",
+      workspace.parentNodeId
+    );
+    xhr.setRequestHeader("X-PFI-Workspace-Name", workspace.workspaceName);
+  }
+  xhr.setRequestHeader("X-PFI-Last-Modified", file.lastModified.toString());
```

Comment:
Would it be worth setting an extra request header here e.g. "X-PFI-Retry-Count"? I think we might be able to pick that up in the request logs from the server so if we get a user complaint we can see whether the retry mechanism was triggered

Reply:
good idea, done